x-pack/plugins/security_solution/server/lib/risk_engine/tasks/risk_scoring_task.ts

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import moment from 'moment';
+import { asyncForEach } from '@kbn/std';
+import {
+  type Logger,
+  SavedObjectsErrorHelpers,
+  type StartServicesAccessor,
+} from '@kbn/core/server';
+import type {
+  ConcreteTaskInstance,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '@kbn/task-manager-plugin/server';
+import { SavedObjectsUtils } from '@kbn/core-saved-objects-utils-server';
+
+import type { AfterKeys, IdentifierType } from '../../../../common/risk_engine';
+import type { StartPlugins } from '../../../plugin';
+import { type RiskScoreService, riskScoreServiceFactory } from '../risk_score_service';
+import { RiskEngineDataClient } from '../risk_engine_data_client';
+import { isRiskScoreCalculationComplete } from '../helpers';
+import {
+  defaultState,
+  stateSchemaByVersion,
+  type LatestTaskStateSchema as RiskScoringTaskState,
+} from './state';
+import { INTERVAL, SCOPE, TIMEOUT, TYPE, VERSION } from './constants';
+import { convertRangeToISO } from './helpers';
+import { RiskScoreEntity } from '../../../../common/risk_engine/types';
+
+const logFactory =
+  (logger: Logger, taskId: string) =>
+  (message: string): void =>
+    logger.info(`[task ${taskId}]: ${message}`);
+
+const getTaskName = (): string => TYPE;
+
+const getTaskId = (namespace: string): string => `${TYPE}:${namespace}:${VERSION}`;
+
+type GetRiskScoreService = (namespace: string) => Promise<RiskScoreService>;
+
+const buildSpaceExtensions = (namespace: string) => ({
+  spacesExtension: {
+    getCurrentNamespace: () => SavedObjectsUtils.namespaceStringToId(namespace),
+    getSearchableNamespaces: async () => [namespace],
+  },
+});
+
+export const registerRiskScoringTask = ({
+  getStartServices,
+  kibanaVersion,
+  logger,
+  taskManager,
+}: {
+  getStartServices: StartServicesAccessor<StartPlugins>;
+  kibanaVersion: string;
+  logger: Logger;
+  taskManager: TaskManagerSetupContract | undefined;
+}): void => {
+  if (!taskManager) {
+    logger.info('Task Manager is unavailable; skipping risk engine task registration.');
+    return;
+  }
+
+  const getRiskScoreService: GetRiskScoreService = (namespace) =>
+    getStartServices().then(([coreStart, _]) => {
+      const esClient = coreStart.elasticsearch.client.asInternalUser;
+      const soClient = coreStart.savedObjects.createInternalRepository(
+        undefined,
+        buildSpaceExtensions(namespace)
+      );
```

Comment:
@elastic/kibana-core can someone please validate that this usage is correct? This task needs to be executed in an arbitrary kibana space, outside of the context of a request. I encode the `namespace` as part of the task's state during scheduling, and then the `soClient` eventually receives that knowledge here. Without [this code](dfdb10ab5749805d105d50190cfe3a90f14b4111), this internal SO client was unable to retrieve a saved object (namely the `risk-engine-configuration`) from a non-default space. It's also possible that the SO was created incorrectly to cause this behavior, but it looked correct `namespaces: ['example-space']` and this code now works.

Reply:
@TinaHeiligers thank you for the review! Would you be able to take a look at this code as well? I couldn't find any examples of people using these extensions, and this was just the first combination of `getCurrentNamespace` and `getSearchableNamespaces` that worked for my purposes.

Reply:
The @elastic/kibana-security team "owns" the extensions. @jeramysoucy can you 👁️ too please?
I'll review this PR.

Reply:
@rylnd the code looks fine to me according to the `jsdocs` for extensions. When it comes to namespace handling, I think that's fine too. I've asked the Kibana security folks to take a look because I think your implementation is the first (or one of the first) outside of Core and Security domains.

Reply:
Thank you @TinaHeiligers 🙏 . One more point that gave me pause here: the extensions docs make a point about "namespace ID" not being the same as a namespace string, and reference the corresponding `SavedObjectUtils`, but they are not explicit about which (strings or IDs) are passed as arguments to these extension functions, nor which are expected to be returned from them.

Reply:
@rylnd Let me make sure I understand the use case. This creates an SO client using a custom spaces extension that limits all operation to the current space. You're not using the security extension or encryption extension. You're using this SO client to create/update the configuration for the risk engine. You rely on the find function to retrieve the config object for the current space. Is this all accurate? Anything missing?

Is there a reason you couldn't use the existing spaces extension? It is designed to operate only within the active space, and you can limit which searched spaces with the find function options.

I will be on PTO next week so I am tagging @azasypkin and @legrego in case you need to follow up on this sooner.

Reply:
@jeramysoucy thanks for the 👀 ! 

> This creates an SO client using a custom spaces extension that limits all operation to the current space

Nope, this client is only used to _retrieve_ the SO; the SO itself is created with the scoped SO client from the request context. The use case is simply "retrieve our configuration SO for the current space," and when I was using the client returned from an empty `createInternalRepository()` call, it was not able to retrieve that SO in a non-default space.

As I mentioned above, the other possibility is that the SO itself is somehow [created](https://github.com/elastic/kibana/pull/163216/files#diff-dde99144e175122a90c2e43b30bd9ea91034597c44e85fa296b78ba9786cf022R90-R94) improperly, but since I was able to find it with this extension that seems less likely? 

If you're saying that this _should_ work without these extensions (is that what you're saying?), I'll dig deeper next week. I would definitely prefer not to use these extensions, and certainly not a subset of them, if possible.

Reply:
@jeramysoucy @azasypkin @legrego I have not had luck using `createInternalRepository` during kibana setup to obtain the SO we need for the particular space at task runtime. Logically this makes sense, since during kibana setup there isn't really a concept of "current space" (or it's simply the default space). 

I can use anything from the `savedObjects` plugin to accomplish this task, but I do not have a `request` object from which to derive a scoped client (since that would require serialization of the request); I only have a namespace string. Is there a better way to accomplish my goal, or are extensions the way to do this? Any assistance would be appreciated.

Reply:
Thanks for all the help here, everyone! After discussion with both @azasypkin and @legrego I opted to go with the (fake) request-based route for instantiating a scoped SO client: [64d9cf5](https://github.com/elastic/kibana/pull/163216/commits/64d9cf59948b0e26b6676ffd1e306b439525e630). 

Things look to be good (as evidenced by smoke testing and a [new integration test](https://github.com/elastic/kibana/pull/163216/commits/c8c1c12a254d152264de70f43a069158a73d51a8)), but please give this a once over and confirm that things look good on your end!

Reply:
Considering that we use the SO client without the security extension only within a background task and solely for a limited and seemingly safe set of operations, looks good to me. I just left one comment/ask.